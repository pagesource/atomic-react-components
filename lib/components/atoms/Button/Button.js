// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';

import { getLabelFromChildren } from '../../../core/utils/elementLocator';
import styles from './Button.style';

type Props = {
  className?: string,
  children: Node,
  inheritedStyles?: string,
  type?: string,
  primary?: boolean,
  secondary?: boolean,
  tertiary?: boolean,
  disabled?: boolean,
  variation?: string,
  transparent?: boolean,
  ariaLabel?: string,
  elementLocator?: string,
};

const Button = ({
  className,
  children,
  inheritedStyles,
  type,
  primary,
  secondary,
  tertiary,
  variation,
  disabled,
  transparent,
  ariaLabel,
  elementLocator,
  ...others
}: Props): Node => (
  <button
    {...others}
    aria-label={ariaLabel || null}
    className={className}
    disabled={disabled}
    type={type}
    data-locator={
      elementLocator || (type ? `button-${type}-${getLabelFromChildren(children)}` : '')
    }
    variation={variation}
  >
    {children}
  </button>
);

Button.defaultProps = {
  inheritedStyles: '',
  type: 'button',
  primary: false,
  secondary: false,
  tertiary: false,
  disabled: false,
  variation: '',
  transparent: false,
  ariaLabel: '',
  className: '',
  elementLocator: '',
};

const StyledButton: ComponentType<Props> = styled(Button)`
  ${styles};
`;
export default StyledButton;

export { Button as ButtonVanilla };
