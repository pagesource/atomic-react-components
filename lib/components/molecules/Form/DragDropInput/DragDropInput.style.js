import { css } from 'styled-components';

// TODO: Fix the duplicate styling in input.style.js
const styles = css`
  width: 100%;
  margin-bottom: ${(props) => props.theme.spacing.UNIT_2};
  min-height: ${(props) => props.theme.spacing.UNIT_20};
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${(props) => props.theme.spacing.UNIT_4};
  ${(props) => `${props.theme.border.DRAG_FILE_FIELD}`};
  ${(props) => `${props.theme.colors.TEXT_GREY}`};
  border-radius: 0.5rem;
  input {
    display: none;
  }
`;

export default styles;
