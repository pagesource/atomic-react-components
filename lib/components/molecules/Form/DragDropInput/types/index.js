// @flow

export type DragDropInputProps = {
  id: string,
  register: Function,
  setValue: Function,
  className?: string,
  name: string,
  label: Node,
  labelProps?: Object,
  fieldProps?: Object,
  localeText: {
    btnTextPrimary: string,
    btnTextSecondary: string,
    dragHelperText?: string,
  },
  onChange?: Function,
  onBlur?: Function,
};
