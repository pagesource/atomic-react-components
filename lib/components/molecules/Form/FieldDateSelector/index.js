// @flow
export { default } from './FieldDateSelector';
export { FieldDateSelectorVanilla } from './FieldDateSelector';
