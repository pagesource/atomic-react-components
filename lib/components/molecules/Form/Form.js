// @flow
import React from 'react';
import { Formik, Form as FormWrapper, Field } from 'formik';
import type { Node } from 'react';
import { buildSchema } from '../../../core/utils/buildValidationSchema';
import { noop } from '../../../core/utils/noop';
import Debug from './Debug';
import FieldInput from './FieldInput';
import FieldSelect from './FieldSelect';
import FieldTextarea from './FieldTextarea';
import FieldRadioGroup from './FieldRadioGroup';
import FieldCheckbox from './FieldCheckbox';
import FieldToggle from './FieldToggle';
import FieldFileInput from './FieldFileInput';
import DDFileInput from './DragDropInput';

type Props = {
  className?: string,
  action: string,
  initialValues?: any,
  children: Node,
  handleSubmit: Function,
  debug?: boolean,
  validationSchema?: Object | null,
  validate?: Function | null,
  validateOnChange?: boolean,
  validateOnBlur?: boolean,
  formWrapperProps?: Object,
  yupValidationSchema?: Function | null,
};

class Form extends React.PureComponent<Props> {
  // eslint-disable-next-line
  static defaultProps = {
    debug: false,
    className: '',
    initialValues: null,
    validationSchema: null,
    validateOnChange: true,
    validateOnBlur: true,
    formWrapperProps: {},
    yupValidationSchema: null,
    validate: noop,
  };

  // eslint-disable-next-line react/sort-comp
  getSchema = (
    yupValidationSchema: Function | null = null,
    validationSchema: Object | null = null
  ) => {
    if (yupValidationSchema) return yupValidationSchema;
    return validationSchema ? buildSchema(validationSchema) : null;
  };

  static Checkbox = FieldCheckbox;

  static Field = Field;

  static Input = FieldInput;

  static RadioGroup = FieldRadioGroup;

  static Select = FieldSelect;

  static Textarea = FieldTextarea;

  static Toggle = FieldToggle;

  static FieldFile = FieldFileInput;

  static DDFile = DDFileInput;

  render() {
    const {
      action,
      className,
      handleSubmit,
      initialValues,
      children,
      debug,
      validationSchema,
      validate,
      validateOnChange,
      validateOnBlur,
      formWrapperProps,
      yupValidationSchema,
      ...others
    } = this.props;
    return (
      <Formik
        initialValues={initialValues}
        validationSchema={this.getSchema(yupValidationSchema, validationSchema)}
        validate={validate}
        onSubmit={handleSubmit}
        validateOnChange={validateOnChange}
        validateOnBlur={validateOnBlur}
        {...formWrapperProps}
      >
        {() => (
          <FormWrapper {...others} noValidate>
            {children}
            {debug && <Debug />}
          </FormWrapper>
        )}
      </Formik>
    );
  }
}

export default Form;
