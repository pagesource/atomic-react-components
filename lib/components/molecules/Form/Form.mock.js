import { action } from '@storybook/addon-actions';

const userFormProps = {
  initialValues: {
    email: '',
    username: '',
    typeOfUser: 'Admin',
    comment: '',
    ageGroup: '',
    termsAndConditions: false,
  },
  debug: false,
  action: '/',
  validationSchema: {
    username: {
      max: [15, 'Max limit reached'],
      required: 'User name Required',
    },
    email: {
      max: [30, 'Max reached'],
      email: 'Invalid email format',
      required: 'Required',
    },
    comment: {
      max: [100, 'Comment should not be more than 100 characters'],
      required: 'Please enter a comment',
    },
    ageGroup: {
      required: 'Please select an Age group',
    },
    termsAndConditions: {
      type: 'bool',
      test: [
        'termsAndConditions',
        'please accept terms and conditions!!!',
        (value) => value === true,
      ],
    },
  },
  onSubmit: action('On Submit Fired'),
};

const userFormSelectProps = {
  options: ['Admin', 'Read Only'],
  name: 'typeOfUser',
  label: 'Type of user',
};

const userFormNameInputProps = {
  name: 'username',
  type: 'text',
  label: 'User Name',
  validationInstructions: 'Enter valid username',
};

const fileMultipleInputProps = {
  name: 'fileuploadMulti',
  localeText: {
    btnTextPrimary: 'Choose Files',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
  multiple: true,
};
const fileSingleInputProps = {
  name: 'fileupload',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};
const fileDDInputProps = {
  name: 'fileuploadDD',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};

const userFormEmailProps = {
  type: 'email',
  name: 'email',
  label: 'Email',
  validationInstructions: 'Enter valid email',
};

const userFormCommentProps = {
  name: 'comment',
  label: 'Comment',
  placeholder: 'Enter your comment here...',
};

const userFormAgeGroupProps = {
  name: 'ageGroup',
  radioGroupArray: [
    {
      value: '30-',
      label: '< 30',
    },
    {
      value: '30-60',
      label: '30-60',
    },
    {
      value: '60+',
      label: '> 60',
    },
  ],
};

const userFormTnCProps = {
  name: 'termsAndConditions',
  label: 'I accept terms & conditions',
};

const userFormToggleProps = {
  name: 'test_1',
  id: 'checkbox_test_1',
  value: 'off',
  label: 'Toggle',
  indicatorOn: 'On',
  indicatorOff: 'Off',
  disabled: false,
  intialToggleState: false,
  displayIndicators: true,
};

export {
  userFormProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormSelectProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormTnCProps,
  userFormToggleProps,
  fileDDInputProps,
};
