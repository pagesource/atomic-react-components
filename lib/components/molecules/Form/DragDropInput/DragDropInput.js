// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import { Field } from 'formik';
import styled from 'styled-components';

import Label from '../../../atoms/Label';
import { ButtonVanilla as Button } from '../../../atoms/Button/Button';
import Input from '../../../atoms/Input';
import useOpenFileDialog from '../../../../hooks/useOpenFileDialog';
import useDragAndDrop from '../../../../hooks/useDragNDrop';
import type { DragDropInputProps } from './types';
import styles from './DragDropInput.style';

const { openFileDialog } = useOpenFileDialog();

const DDFieldInput = ({
  id,
  className,
  name,
  label,
  onChange,
  onBlur,
  labelProps,
  fieldProps,
  localeText,
  ...others
}: DragDropInputProps): Node => {
  const fileInput = React.createRef();
  const { onDragOver, onDragLeave, onDrop, onFilesAdded, getHighlightClass } = useDragAndDrop();
  const styleCls = getHighlightClass(className);

  return (
    <Button
      className={styleCls}
      onClick={() => openFileDialog(fileInput)}
      onDragOver={onDragOver}
      onDragLeave={onDragLeave}
      onDrop={onDrop}
    >
      <Label htmlFor={id} {...labelProps}>
        {label}
      </Label>
      <Field
        id={id}
        name={name}
        render={({ field, form }) => {
          const { handleChange, handleBlur } = form;
          return (
            <Input
              type="file"
              ref={fileInput}
              {...others}
              {...field}
              onChange={(e) => {
                handleChange(e);
                onFilesAdded(e);
                if (onChange) onChange(e, form);
              }}
              onBlur={(e) => {
                handleBlur(e);
                if (onBlur) onBlur(e, form);
              }}
            />
          );
        }}
        {...fieldProps}
      />
      {localeText.dragHelperText}
    </Button>
  );
};

DDFieldInput.defaultProps = {
  className: 'dragAndDrop',
  labelProps: {},
  fieldProps: {},
  onChange: null,
  onBlur: null,
  localeText: {
    dragHelperText: 'Drop Files Here...',
  },
};

const StyledDDFieldInput: ComponentType<DragDropInputProps> = styled(DDFieldInput)`
  ${styles};
`;

export default StyledDDFieldInput;

export { DDFieldInput as DDFieldInputVanilla };
