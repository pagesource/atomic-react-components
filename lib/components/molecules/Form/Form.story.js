// @flow
import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  userFormProps,
  userFormSelectProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormDateSelectorProps,
  userFormTnCProps,
  userFormToggleProps,
} from './Form.mock';

// Import Vanilla Component to display markup and props
import Form from './Form';
import Button from '../../atoms/Button';
// Import Styled Component to showcase variations

storiesOf('Molecules/Form', module)
  .addParameters({ jest: ['Form', 'Button'] })
  .add('Default', () => (
    <Form {...userFormProps}>
      <div>
        <Form.Select {...userFormSelectProps} />
        <Form.Input {...userFormNameInputProps} />
        <Form.Input {...userFormEmailProps} />
        <Form.FieldFile {...fileMultipleInputProps} />
        <Form.FieldFile {...fileSingleInputProps} />
        <Form.DDFile {...fileSingleInputProps} />
        <Form.RadioGroup {...userFormAgeGroupProps} />
        <Form.DateSelector {...userFormDateSelectorProps} />
        <Form.Checkbox {...userFormTnCProps} />
        <Form.Textarea {...userFormCommentProps} />
        <Form.Toggle {...userFormToggleProps} />
      </div>
      <Button primary type="submit">
        Submit
      </Button>
    </Form>
  ));
