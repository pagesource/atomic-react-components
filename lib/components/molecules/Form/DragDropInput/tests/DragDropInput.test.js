import React from 'react';

import { shallowWithTheme } from '../../../../../../__mock__/themeMock';
import Form from '../../Form';
import { DDFieldInputVanilla } from '../DragDropInput';
import useDragAndDrop from '../../../../../hooks/useDragNDrop';
import { ButtonVanilla } from '../../../../atoms/Button';
import Input from '../../../../atoms/Input';

const defaultProps = {
  name: 'fileupload',
  localeText: {
    btnTextPrimary: 'Choose File(s)',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
};

const defaultPropsTwo = {
  name: 'fileuploadTwo',
  localeText: {
    btnTextPrimary: 'Choose File(s)',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
};

const getWrapperSetup = (props = {}) => {
  const finalProps = { ...defaultProps, ...props };
  return shallowWithTheme(<DDFieldInputVanilla {...finalProps} />);
};

describe('DDFieldInput Component', () => {
  let ddField = '';
  beforeEach(() => {
    ddField = getWrapperSetup();
  });

  test('should render correctly', () => {
    ddField = shallowWithTheme(
      <Form>
        <div>
          <DDFieldInputVanilla {...defaultProps} />
          <DDFieldInputVanilla {...defaultPropsTwo} />
        </div>
        <ButtonVanilla primary type="submit">
          Submit
        </ButtonVanilla>
      </Form>
    );
    expect(ddField).toMatchSnapshot();
  });

  test('should render Input component with proper type for multiple file upload.', () => {
    const fileMultipleInputProps = {
      name: 'fileupload',
      multiple: true,
    };
    ddField = getWrapperSetup(fileMultipleInputProps);
    expect(ddField.find('Button')).toHaveLength(1);
  });

  test('should render the component with prop correctly', () => {
    expect(ddField.find('Button').prop('onClick')).toBeTruthy();
  });

  test.skip('should match the input rendered correctly on change event simulation.', () => {
    const inputElement = ddField.find('Input');
    const wrapper = inputElement.simulate('change');
    expect(wrapper.find('Input').length).toBe(1);
  });

  test('should match the input rendered correctly on blur event simulation.', () => {
    const inputElement = ddField.find(Input);
    const wrapper = inputElement.simulate('blur');
    expect(wrapper.find(Input).length).toBe(1);
  });
});

describe('testing the custom hook methods for drag and drop', () => {
  let container = '';

  beforeEach(() => {
    const Elements = () => {
      const props = useDragAndDrop({});
      return <div {...props} />;
    };
    container = shallowWithTheme(<Elements />);
  });

  it('should simulate file upload and return array', () => {
    expect.objectContaining(
      container.props().fileListToArray({
        filesArr: [
          {
            lastModified: 1575972428087,
            lastModifiedDate: 11,
            name: 'index.html',
            size: 3741,
            type: 'text/html',
            webkitRelativePath: '',
          },
        ],
        length: 1,
        item: function () {
          return this.filesArr[0];
        },
      })
    );
  });

  it('should make highlight as true on drag over as mouse enters draggable area', () => {
    expect.objectContaining(
      container.props().onDragOver({
        preventDefault: function () {
          return null;
        },
      })
    );
  });

  it('should make highlight as false on drag leave as mouse leaves draggable area', () => {
    expect.objectContaining(
      container.props().onDragLeave({
        preventDefault: function () {
          return null;
        },
      })
    );
  });

  it('should test the files addition handler when file(s) are uploaded.', () => {
    expect.objectContaining(
      container.props().onFilesAdded({
        target: {
          files: {
            filesArr: [
              {
                lastModified: 1575972428087,
                lastModifiedDate: 11,
                name: 'index.html',
                size: 3741,
                type: 'text/html',
                webkitRelativePath: '',
              },
            ],
            length: 1,
            item: function () {
              return this.filesArr[0];
            },
          },
        },
      })
    );
  });

  it('should apply highlight when cursor moves draggable area.', () => {
    container.props().setHighlight(true);
    expect(container.props().getHighlightClass('container-draggable')).toBe(
      'container-draggable highlight'
    );
  });

  it('should apply highlight when cursor moves draggable area.', () => {
    expect(container.props().getHighlightClass('container-draggable')).toBe('container-draggable');
  });
});
