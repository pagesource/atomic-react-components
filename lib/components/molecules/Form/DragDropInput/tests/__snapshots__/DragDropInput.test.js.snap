// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DDFieldInput Component should render correctly 1`] = `
<Form
  className=""
  debug={false}
  formWrapperProps={Object {}}
  initialValues={null}
  validate={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={null}
  yupValidationSchema={null}
>
  <Formik
    initialValues={null}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
    validationSchema={null}
  >
    <Form
      noValidate={true}
    >
      <form
        action="#"
        noValidate={true}
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <DDFieldInput
          className="dragAndDrop"
          fieldProps={Object {}}
          labelProps={Object {}}
          localeText={
            Object {
              "btnTextPrimary": "Choose File(s)",
              "btnTextSecondary": "Upload...",
              "dragHelperText": "Drop Files Here...",
            }
          }
          name="fileupload"
          onBlur={null}
          onChange={null}
        >
          <Button
            ariaLabel=""
            className="dragAndDrop"
            disabled={false}
            elementLocator=""
            inheritedStyles=""
            onClick={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            primary={false}
            secondary={false}
            tertiary={false}
            transparent={false}
            type="button"
            variation=""
          >
            <button
              aria-label={null}
              className="dragAndDrop"
              data-locator="button-button-"
              disabled={false}
              onClick={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              type="button"
              variation=""
            >
              <Label
                className=""
                elementLocator=""
              >
                <label
                  className=""
                  data-locator="label-undefined-"
                />
              </Label>
              <Field
                name="fileupload"
                render={[Function]}
              >
                <Input__StyledInput
                  name="fileupload"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="file"
                  value={null}
                >
                  <ForwardRef
                    ariaLabel=""
                    className="Input__StyledInput-sc-133uxg0-0 FRPZs"
                    elementLocator=""
                    name="fileupload"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder=""
                    type="file"
                    value={null}
                  >
                    <input
                      aria-label=""
                      className="Input__StyledInput-sc-133uxg0-0 FRPZs"
                      data-locator="input-file-fileupload"
                      name="fileupload"
                      onBlur={[Function]}
                      onChange={[Function]}
                      placeholder=""
                      type="file"
                    />
                  </ForwardRef>
                </Input__StyledInput>
              </Field>
              Drop Files Here...
            </button>
          </Button>
        </DDFieldInput>
      </form>
    </Form>
  </Formik>
</Form>
`;
